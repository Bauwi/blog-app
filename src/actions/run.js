import database from '../firebase/firebase';
import { db } from '../firebase/firebase';
import * as firebase from 'firebase';

export const addPostToRun = post => ({
  type: 'ADD_POST_TO_RUN',
  post
});

export const startAddPostToRun = post => (dispatch, getState) => {
  const { uid } = getState().auth;
  console.log(post);
  db
    .collection('runs')
    .doc(uid)
    .collection('UserRuns')
    .add(post)
    .then((doc) => {
      dispatch(addPostToRun({
        DBid: doc.id,
        ...post
      }));
    });
};

export const removePostToRun = id => ({
  type: 'REMOVE_POST_TO_RUN',
  id
});

export const startRemovePostToRun = (id, DBid) => (dispatch, getState) => {
  const { uid } = getState().auth;
  console.log(DBid);
  return db
    .collection('runs')
    .doc(uid)
    .collection('UserRuns')
    .doc(`${DBid}`)
    .delete()
    .then(dispatch(removePostToRun(id)));
};

export const setRunPosts = posts => ({
  type: 'SET_RUN_POSTS',
  posts
});

export const startSetRunPosts = () => (dispatch, getState) => {
  const { uid } = getState().auth;
  return db
    .collection('runs')
    .doc(uid)
    .collection('UserRuns')
    .get()
    .then((querySnapshot) => {
      const run = [];
      querySnapshot.forEach((doc) => {
        run.push({
          DBid: doc.id,
          ...doc.data()
        });
      });
      dispatch(setRunPosts(run));
    });
};
