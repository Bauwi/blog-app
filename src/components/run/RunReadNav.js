import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';

import RunReadNavItem from './RunReadNavItem';

const RunReadNav = ({ previousRead, nextRead }) => (
  <nav className="post-nav content-container">
    <ul>
      <RunReadNavItem post={previousRead} />
      <RunReadNavItem post={nextRead} />
    </ul>
  </nav>
);

RunReadNav.propTypes = {
  previousRead: PropTypes.object,
  nextRead: PropTypes.object
};

const mapStateToProps = (state, props) => {
  const currentPostIndex = state.run.posts.findIndex(post => post.content.id === props.currentRead.content.id);
  return {
    previousRead: state.run.posts[currentPostIndex - 1],
    nextRead: state.run.posts[currentPostIndex + 1]
  };
};

export { RunReadNav };
export default connect(mapStateToProps)(RunReadNav);
