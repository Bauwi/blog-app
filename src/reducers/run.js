const runReducerDefaultState = {
  isLoading: true, posts: [], hasErrored: false, current: {}
};

const runReducer = (state = runReducerDefaultState, action) => {
  switch (action.type) {
    case 'ADD_POST_TO_RUN':
      return {
        ...state,
        posts: [...state.posts, action.post]
      };
    case 'REMOVE_POST_TO_RUN':
      return {
        ...state,
        posts: state.posts.filter(post => action.id !== post.content.id)
      };
    case 'SET_RUN_POSTS':
      return {
        ...state,
        posts: action.posts
      };
    case 'RUN_IS_LOADING':
      return {
        ...state,
        isLoading: action.isLoading
      };
    case 'RUN_HAS_ERRORED':
      return {
        ...state,
        hasErrored: action.hasErrored
      };
    case 'SET_CURRENT_POST_RUN':
      return {
        ...state,
        current: state.posts.find(post => post.content.id === action.id)
      };
    case 'UPDATE_RUN_POST_TO_ALREADY_READ':
      return {
        ...state,
        current: { ...state.current, state: 'read' },
        posts: state.posts.map((post) => {
          if (post.content.id === action.id) {
            return {
              ...post,
              state: 'read'
            };
          }
          return post;
        })
      };
    case 'UPDATE_RUN_POST_TO_NOT_ALREADY_READ':
      return {
        ...state,
        posts: state.posts.map((post) => {
          if (post.content.id === action.id) {
            return {
              ...post,
              state: 'unread'
            };
          }
          return post;
        })
      };
    case 'RESET_RUN':
      return {
        posts: []
      };
    case 'CLEAN_RUN':
      return {
        ...state,
        posts: state.posts.filter(post => post.state === 'unread')
      };
    default:
      return state;
  }
};

export default runReducer;
