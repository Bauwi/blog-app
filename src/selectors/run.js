const getRunPosts = (posts = [], run) => run.map(read => posts.find(post => post.id === read.id));
export default getRunPosts;
