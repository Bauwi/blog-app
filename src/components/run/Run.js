import React, { Component } from 'react';
import { connect } from 'react-redux';

import InlineList from '../InlineList';
import LoadingPage from '../LoadingPage';
import { startSetRunPosts } from '../../actions/run';

export class Run extends Component {
  state = {
    loading: true
  };
  componentWillMount() {
    if (this.props.run.length === 0) {
      this.props.startSetRunPosts().then(() => {
        this.setState(() => ({ loading: false }));
      });
    } else {
      this.setState(() => ({ loading: false }));
    }
  }

  render() {
    if (this.state.loading) {
      return <LoadingPage />;
    }
    return (
      <div>
        <p>Run</p>
        <InlineList posts={this.props.run} />
      </div>
    );
  }
}

const mapDispatchToProps = dispatch => ({
  startSetRunPosts: () => dispatch(startSetRunPosts())
});

const mapStateToProps = state => ({
  run: state.run
});

export default connect(mapStateToProps, mapDispatchToProps)(Run);
