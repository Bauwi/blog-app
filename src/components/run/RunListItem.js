import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { withRouter } from 'react-router-dom';
import moment from 'moment';

import KeywordsList from '../KeywordsList';

import { setCurrentPostRun } from '../../actions/run';

export class RunListItem extends Component {
  handleItemClick = () => {
    this.props.history.push('/run/start');
    this.props.setCurrentPostRun(this.props.post.content.id);
  };

  render() {
    const {
      id,
      title,
      createdAt,
      author,
      keywords,
      authorId,
      miniCover,
      stars
    } = this.props.post.content;
    return (
      <div>
        <div className="inline-list-item">
          <header className="inline-list-item__left">
            <img src={miniCover} className="inline-list-item__cover" alt="post cover" />
          </header>
          <div className="inline-list-item__right">
            <section className="inline-list-item__header">
              <div>
                <i className="fa fa-star" /> {stars}
              </div>
              <div className="inline-list-item__header--right">
                {this.props.state === 'read' && (
                  <p>
                    <i className="fa fa-check" />
                  </p>
                )}
                <p className="inline-list-item__date">{moment(createdAt).format('MMM Do, YYYY')}</p>
              </div>
            </section>
            <div onClick={this.handleItemClick} className="inline-list-item__infos">
              <h3 className="inline-list-item__title">{title}</h3>
            </div>
            <footer className="inline-list-item__footer">
              <p className="inline-list-item__author">{author}</p>
              <KeywordsList keywords={keywords} />
            </footer>
          </div>
        </div>
      </div>
    );
  }
}

RunListItem.propTypes = {
  post: PropTypes.object.isRequired,
  state: PropTypes.string.isRequired,
  setCurrentPostRun: PropTypes.func.isRequired
};

const mapDispatchToProps = dispatch => ({
  setCurrentPostRun: id => dispatch(setCurrentPostRun(id))
});

export default withRouter(connect(undefined, mapDispatchToProps)(RunListItem));
