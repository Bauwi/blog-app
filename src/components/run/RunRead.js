import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { withRouter } from 'react-router-dom';
import { connect } from 'react-redux';
import { notification } from 'antd';

import RunReadNav from './RunReadNav';
import RunReadPost from './RunReadPost';
import LoadingPage from '../LoadingPage';
import { startUpdateRunPostToAlreadyRead } from '../../actions/run';

export class RunRead extends Component {
  componentDidMount() {
    window.addEventListener('scroll', this.handleScroll);
  }

  componentWillUnmount() {
    window.removeEventListener('scroll', this.handleScroll);
  }

  handleScroll = event => {
    const DBid = this.props.current.DBid;
    if (window.innerHeight + window.scrollY >= document.body.scrollHeight) {
      if (this.props.current.state !== 'read') {
        const { id, title, author } = this.props.current.content;
        const openNotificationWithIcon = type => {
          notification[type]({
            message: title,
            description: `by ${author}`,
            style: {
              position: 'relative',
              top: '4rem'
            }
          });
        };
        openNotificationWithIcon('success');
        this.props.startUpdateRunPostToAlreadyRead(id, DBid);
      }
    }
  };

  render() {
    if (this.props.isLoading) {
      return <LoadingPage />;
    }

    return (
      <div>
        <RunReadPost post={this.props.current} />
        <RunReadNav currentRead={this.props.current} />
      </div>
    );
  }
}

RunRead.propTypes = {
  current: PropTypes.object,
  isLoading: PropTypes.bool.isRequired,
  startUpdateRunPostToAlreadyRead: PropTypes.func.isRequired
};

const mapDispatchToProps = dispatch => ({
  startUpdateRunPostToAlreadyRead: (id, DBid) => dispatch(startUpdateRunPostToAlreadyRead(id, DBid))
});

const mapStateToProps = state => ({
  current: state.run.current,
  isLoading: state.run.isLoading
});

export default withRouter(connect(mapStateToProps, mapDispatchToProps)(RunRead));
