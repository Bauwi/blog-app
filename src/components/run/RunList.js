import React, { Component } from 'react';

import RunListItem from './RunListItem';

export default class RunList extends Component {
  renderList() {
    return this.props.posts.map((post) => {
      const key = post.content ? post.content.id : post.id;

      return <RunListItem key={key} post={post} state={post.state} />;
    });
  }

  render() {
    return (
      <div className="content-container run-list">
        <p>RunList</p>
        {this.props.posts ? (
          this.renderList()
        ) : (
          <div>Add a post to your run to start a session</div>
        )}
      </div>
    );
  }
}
