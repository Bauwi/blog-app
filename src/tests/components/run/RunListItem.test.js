import React from 'react';
import { shallow } from 'enzyme';
import { RunListItem } from '../../../components/run/RunListItem';
import run from '../../fixtures/run';

test('should render RunListItem properly if not read yet', () => {
  const wrapper = shallow(<RunListItem state={run.posts[1].state} post={run.posts[1]} setCurrentPostRun={() => {}} />);
  expect(wrapper).toMatchSnapshot();
});

test('should render RunListItem properly if already read', () => {
  const wrapper = shallow(<RunListItem state={run.posts[0].state} post={run.posts[0]} setCurrentPostRun={() => {}} />);
  expect(wrapper).toMatchSnapshot();
});

test('should call setCurrentPostRun and redirect if clicked', () => {
  const setCurrentPostRun = jest.fn();
  const history = { push: jest.fn() };
  const wrapper = shallow(<RunListItem
    state={run.posts[0].state}
    post={run.posts[0]}
    history={history}
    setCurrentPostRun={setCurrentPostRun}
  />);
  wrapper.find('.inline-list-item__infos').simulate('click');
  expect(history.push).toHaveBeenLastCalledWith('/run/start');
  expect(setCurrentPostRun).toHaveBeenLastCalledWith(run.posts[0].content.id);
});
