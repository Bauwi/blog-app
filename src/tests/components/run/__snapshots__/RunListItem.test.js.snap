// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render RunListItem properly if already read 1`] = `
<div
  onClick={[Function]}
>
  <div
    className="inline-list-item"
  >
    <header
      className="inline-list-item__left"
    >
      <img
        alt="post cover"
        className="inline-list-item__cover"
        src="https://www.w3schools.com/w3css/img_fjords.jpg"
      />
    </header>
    <div
      className="inline-list-item__right"
    >
      <section
        className="inline-list-item__header"
      >
        <div>
          <i
            className="fa fa-star"
          />
           
          5
        </div>
        <div
          className="inline-list-item__header--right"
        >
          <p>
            <i
              className="fa fa-check"
            />
          </p>
          <p
            className="inline-list-item__date"
          >
            Jan 1st, 1970
          </p>
        </div>
      </section>
      <div
        className="inline-list-item__infos"
      >
        <h3
          className="inline-list-item__title"
        >
          Post 1 in Run
        </h3>
      </div>
      <footer
        className="inline-list-item__footer"
      >
        <p
          className="inline-list-item__author"
        >
          author1
        </p>
        <KeywordsList
          keywords="red, green, blue"
        />
      </footer>
    </div>
  </div>
</div>
`;

exports[`should render RunListItem properly if not read yet 1`] = `
<div
  onClick={[Function]}
>
  <div
    className="inline-list-item"
  >
    <header
      className="inline-list-item__left"
    >
      <img
        alt="post cover"
        className="inline-list-item__cover"
        src="https://www.w3schools.com/w3css/img_fjords.jpg"
      />
    </header>
    <div
      className="inline-list-item__right"
    >
      <section
        className="inline-list-item__header"
      >
        <div>
          <i
            className="fa fa-star"
          />
           
          25
        </div>
        <div
          className="inline-list-item__header--right"
        >
          <p
            className="inline-list-item__date"
          >
            Jan 1st, 1970
          </p>
        </div>
      </section>
      <div
        className="inline-list-item__infos"
      >
        <h3
          className="inline-list-item__title"
        >
          Post 2 in Run
        </h3>
      </div>
      <footer
        className="inline-list-item__footer"
      >
        <p
          className="inline-list-item__author"
        >
          author2
        </p>
        <KeywordsList
          keywords="cat, dog, fish"
        />
      </footer>
    </div>
  </div>
</div>
`;
